File: sdks/python/layer5/integrations/openai.py
"""
OpenAI SDK integration for Layer5.
Covers: function calling, tool use, Assistants API tool runs.

Supports: openai >= 1.0.0

Usage — function calling wrapper:
    from openai import OpenAI
    from layer5.integrations.openai import Layer5OpenAIWrapper

    openai_client = OpenAI()
    l5_client     = Layer5(api_key="layer5_...")

    wrapper = Layer5OpenAIWrapper(
        openai_client=openai_client,
        layer5_client=l5_client,
        agent_id="openai-agent"
    )

    response = wrapper.chat_completion(
        model="gpt-4o",
        messages=messages,
        tools=tools,
    )

Usage — manual tool tracking:
    from layer5.integrations.openai import track_tool_calls

    response = openai_client.chat.completions.create(
        model="gpt-4o",
        messages=messages,
        tools=tools
    )

    track_tool_calls(
        layer5_client=l5_client,
        agent_id="openai-agent",
        tool_calls=response.choices[0].message.tool_calls,
        results=your_execution_results,
    )
"""

from __future__ import annotations

import json
import warnings
from datetime import datetime
from typing import Any, Callable, Dict, List, Optional

from ..client import Layer5
from ..exceptions import Layer5Error


def _require_openai() -> Any:
    try:
        import openai

        return openai
    except ImportError:
        raise ImportError(
            "OpenAI SDK is not installed. "
            "Install it: pip install openai\n"
            "Or: pip install layer5[openai]"
        )


def track_tool_calls(
    layer5_client: Layer5,
    agent_id: str,
    tool_calls: List[Any],
    results: List[Dict[str, Any]],
    silent_errors: bool = True,
    decision_ids: Optional[List[Optional[str]]] = None,
) -> None:
    """
    Log outcomes for OpenAI tool calls you executed.

    Call this AFTER you execute all tool calls from
    an OpenAI response and have the results.

    Args:
        layer5_client: Initialized Layer5 client
        agent_id:      Your agent identifier
        tool_calls:    response.choices[0].message.tool_calls
        results:       List of dicts, one per tool call:
                       [{"success": True,
                         "response_ms": 241,
                         "outcome_score": 0.9},  # optional
                        ...]
                       Length must match tool_calls.
        silent_errors: True = don't crash on Layer5 errors
    """
    if len(tool_calls) != len(results):
        raise ValueError(
            f"tool_calls length ({len(tool_calls)}) must "
            f"match results length ({len(results)}). "
            f"Provide one result dict per tool call."
        )

    for i, (tool_call, result) in enumerate(zip(tool_calls, results)):
        func_name = tool_call.function.name

        try:
            args = json.loads(tool_call.function.arguments or "{}")
        except Exception:
            args = {}

        decision_id = (
            decision_ids[i] if decision_ids and i < len(decision_ids) else None
        )

        try:
            layer5_client.log_outcome(
                agent_id=agent_id,
                action_name=func_name,
                session_id="openai-auto",
                issue_type=func_name,
                success=result.get("success", True),
                raw_context={"tool": func_name, **args},
                response_time_ms=result.get("response_ms"),
                outcome_score=result.get("outcome_score"),
                business_outcome=result.get("business_outcome"),
                decision_id=decision_id,
            )
        except Layer5Error as e:
            if not silent_errors:
                raise
            warnings.warn(
                f"[Layer5] log_outcome for '{func_name}' failed: {e}",
                stacklevel=2,
            )


class Layer5OpenAIWrapper:
    """
    Drop-in wrapper around OpenAI client that adds
    Layer5 tracking to all tool calls automatically.
    """

    def __init__(
        self,
        openai_client: Any,
        layer5_client: Layer5,
        agent_id: str,
        tool_executor: Optional[Callable[..., Dict[str, Any]]] = None,
        silent_errors: bool = True,
    ) -> None:
        """
        Args:
            openai_client: Your initialized OpenAI client
            layer5_client: Your initialized Layer5 client
            agent_id:      Agent identifier
            tool_executor: Optional callable to auto-execute tool calls.
                          fn(tool_call) -> {"success": bool, "response_ms": int}
                          If None: tracks from response only.
            silent_errors: Layer5 errors never crash your code
        """
        _require_openai()
        self._openai = openai_client
        self._layer5 = layer5_client
        self.agent_id = agent_id
        self.tool_executor = tool_executor
        self.silent_errors = silent_errors

    def chat_completion(
        self,
        model: str,
        messages: List[Dict[str, Any]],
        tools: Optional[List[Any]] = None,
        **kwargs: Any,
    ) -> Any:
        """
        Drop-in for openai_client.chat.completions.create().
        Automatically tracks tool call outcomes if
        tool_executor is provided.

        Returns: same OpenAI ChatCompletion response object.
        """
        response = self._openai.chat.completions.create(
            model=model,
            messages=messages,
            tools=tools,
            **kwargs,
        )

        message = response.choices[0].message
        if self.tool_executor and message.tool_calls:
            results: List[Dict[str, Any]] = []
            for tc in message.tool_calls:
                start = datetime.now()
                try:
                    self.tool_executor(tc)
                    results.append(
                        {
                            "success": True,
                            "response_ms": int(
                                (datetime.now() - start).total_seconds() * 1000
                            ),
                        }
                    )
                except Exception:
                    results.append(
                        {
                            "success": False,
                            "response_ms": int(
                                (datetime.now() - start).total_seconds() * 1000
                            ),
                        }
                    )

            track_tool_calls(
                layer5_client=self._layer5,
                agent_id=self.agent_id,
                tool_calls=message.tool_calls,
                results=results,
                silent_errors=self.silent_errors,
            )

        return response

File: sdks/typescript/src/integrations/openai.ts
/**
 * OpenAI TypeScript SDK integration.
 * Covers: chat completions with tool calls,
 *         Assistants API function calls.
 *
 * Works with: openai >= 4.0.0
 *
 * @example
 * ```ts
 * import OpenAI from 'openai';
 * import { Layer5 } from '@layer5/sdk';
 * import { trackToolCalls } from '@layer5/sdk/integrations/openai';
 *
 * const openai = new OpenAI();
 * const l5     = new Layer5({ apiKey: 'layer5_...' });
 *
 * const response = await openai.chat.completions.create({
 *   model: 'gpt-4o',
 *   messages,
 *   tools,
 * });
 *
 * const toolCalls = response.choices[0].message.tool_calls ?? [];
 * const results = await Promise.all(
 *   toolCalls.map(async (tc) => {
 *     const start = Date.now();
 *     try {
 *       await executeToolCall(tc);
 *       return { success: true, responseMs: Date.now() - start };
 *     } catch {
 *       return { success: false, responseMs: Date.now() - start };
 *     }
 *   })
 * );
 *
 * await trackToolCalls({ client: l5, agentId: 'openai-agent', toolCalls, results });
 * ```
 */

import type { Layer5 } from '../client.js';

interface ToolCallResult {
  success: boolean;
  responseMs?: number;
  outcomeScore?: number;
  businessOutcome?: string;
}

interface OpenAIToolCall {
  id: string;
  type: 'function';
  function: {
    name: string;
    arguments: string;
  };
}

export async function trackToolCalls(options: {
  client: Layer5;
  agentId: string;
  toolCalls: OpenAIToolCall[];
  results: ToolCallResult[];
  silentErrors?: boolean;
  /** Optional decision IDs — one per tool call. */
  decisionIds?: Array<string | null | undefined>;
}): Promise<void> {
  const {
    client,
    agentId,
    toolCalls,
    results,
    silentErrors = true,
    decisionIds,
  } = options;

  if (toolCalls.length !== results.length) {
    throw new Error(
      `trackToolCalls: toolCalls length (${toolCalls.length}) ` +
      `must match results length (${results.length}). ` +
      `Provide one result object per tool call.`
    );
  }

  await Promise.allSettled(
    toolCalls.map(async (tc, i) => {
      const result = results[i]!;
      let args: Record<string, unknown> = {};

      try {
        args = JSON.parse(tc.function.arguments || '{}');
      } catch {
        // ignore parse errors
      }

      try {
        await client.logOutcome({
          agentId,
          actionName: tc.function.name,
          success: result.success,
          context: { tool: tc.function.name, ...args },
          responseMs: result.responseMs,
          outcomeScore: result.outcomeScore,
          businessOutcome: result.businessOutcome as 'resolved' | 'partial' | 'failed' | 'unknown' | undefined,
          decisionId: decisionIds?.[i] ?? undefined,
        });
      } catch (e) {
        if (!silentErrors) throw e;
        console.warn(
          `[Layer5] logOutcome failed for ` +
          `'${tc.function.name}': ${(e as Error).message}`
        );
      }
    })
  );
}

export function withLayer5<T extends {
  chat: {
    completions: {
      create: (...args: unknown[]) => Promise<unknown>
    }
  }
}>(
  openaiClient: T,
  layer5Client: Layer5,
  options: {
    agentId: string;
    toolExecutor?: (
      toolCall: OpenAIToolCall
    ) => Promise<ToolCallResult>;
    silentErrors?: boolean;
  }
): T {
  const { agentId, toolExecutor, silentErrors = true } = options;

  return new Proxy(openaiClient, {
    get(target, prop) {
      if (prop !== 'chat') return (target as Record<string | symbol, unknown>)[prop];

      return new Proxy(target.chat, {
        get(chatTarget, chatProp) {
          if (chatProp !== 'completions')
            return (chatTarget as Record<string | symbol, unknown>)[chatProp];

          return new Proxy(chatTarget.completions, {
            get(compTarget, compProp) {
              if (compProp !== 'create')
                return (compTarget as Record<string | symbol, unknown>)[compProp];

              return async (...args: unknown[]) => {
                const response = await (
                  compTarget.create as (...a: unknown[]) => Promise<unknown>
                )(...args);

                const message =
                  (response as {
                    choices: Array<{
                      message: {
                        tool_calls?: OpenAIToolCall[]
                      }
                    }>
                  }).choices[0]?.message;

                if (
                  toolExecutor &&
                  message?.tool_calls?.length
                ) {
                  const results = await Promise.all(
                    message.tool_calls.map(tc =>
                      toolExecutor(tc).catch(() => ({
                        success: false,
                        responseMs: 0,
                      }))
                    )
                  );

                  try {
                    await trackToolCalls({
                      client: layer5Client,
                      agentId,
                      toolCalls: message.tool_calls,
                      results,
                      silentErrors,
                    });
                  } catch (e) {
                    if (!silentErrors) throw e;
                    console.warn(
                      `[Layer5] trackToolCalls failed: ${(e as Error).message}`
                    );
                  }
                }

                return response;
              };
            },
          });
        },
      });
    },
  }) as T;
}

export type { OpenAIToolCall, ToolCallResult };
